Add ColorBar.display_low and display_high

ianthomas23 opened the pull request:
- [x] issues: partially fixes #12183
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

This PR adds `low_cutoff` and `high_cutoff` to `ColorBar`. It supersedes PR #12184 and adds the new functionality required to fix issue #12183 but the actual fix will follow in a separate PR that will programmatically set the `low_cutoff` as required.

Note that only the color bar output is cutoff, data that is colored in a plot using a `ColorMapper` that is the subject of `ColorBar` cutoffs is unaffected and will use the whole color range (unless requested not to in the usual way).

Demo:

![cutoff](https://user-images.githubusercontent.com/580326/177162594-b02920c4-999d-4373-b4bf-38040e8fc7f4.gif)

using code
```python
from bokeh.io import show
from bokeh.layouts import column
from bokeh.models import ColorBar, ColumnDataSource, Slider, LinearColorMapper
from bokeh.palettes import Spectral10
from bokeh.plotting import figure
import numpy as np

p = figure(width=500, height=300)
color_mapper = LinearColorMapper(Spectral10)

x = np.arange(1, 11)
cds = ColumnDataSource(data=dict(x=x, y=x, v=x))
c = p.circle(source=cds, x="x", y="y", size=40, fill_color=dict(field="v", transform=color_mapper))

color_bar = ColorBar(color_mapper=color_mapper, bar_line_color="black")
p.add_layout(color_bar, "right")

low_slider = Slider(start=-1, end=10, step=0.1, value=-1, title="low cutoff")
low_slider.js_link("value", color_bar, "low_cutoff")

high_slider = Slider(start=-1, end=10, step=0.1, value=10, title="high cutoff")
high_slider.js_link("value", color_bar, "high_cutoff")

show(column(p, low_slider, high_slider))
```
The whole of the colored rectangle corresponding to the cutoff value is displayed, so there won't just be a single row of colored pixels for example. If `high_cutoff < low_cutoff` or `high_cutoff` is lower than the minimum data value or `low_cutoff` is higher than the maximum data value then no colors are displayed.

Implementation details:

1. There is separate handling of EqHist and Linear/Log color mapper classes in `ColorBar`, as was the case before this PR.
2. `ColorBar._image` can now be `null` if there is no image to display.
3. There are extra tweaks in some `ColorMapper`, `Ticker` and `Axis` classes to deal with empty data.
4. The `ContinuousColorMapper`-derived classes now have separate functions for transforms from data value to palette indices and back again. Some of the code for this was originally in the `cmap` functions. The `cmap` function is now solely in `ContinuousColorMapper` as it doesn't need to be overridden. It also no longer takes the `scan_data` as an argument as it was always recalculated and stored as `this._scan_data` before `cmap` was called, so I am just reading it from `this._scan_data`.

(Edited to tick release notes check box)

ianthomas23 commented:
I need to rebase this and deal with the conflicts in the colorbar code caused by the contouring PR.

ianthomas23 commented:
I have rebased this.

bryevdv commented:
The failure is for bundle size 
```
✗ js/bokeh.min.js          851 kB >   850 kB
```

ianthomas23 commented:
> The failure is for bundle size

I've increase the bundle size from 850 to 900 kB as it looks like we have historically used 50 kB increments.

bryevdv commented on bokeh/models/annotations/legends.py:
```diff
@@ -234,6 +234,16 @@ def __init__(self, *args, **kwargs) -> None:
+    high_cutoff = Nullable(Float, help="""
+    The highest value to show in the color bar. The whole of the color entry
+    containing this value is shown.
+    """)
```
Would it make sense to use names like `display_low` and `display_high` (or similar) to help reinforce that these properties only affect the visual presentation, and not the color mapping?

ianthomas23 replied:
Yes, those are better names. I've also changed `low_index` and `high_index` to `index_low` and `index_high` to better associate them with `display_low` and `display_high`.

ianthomas23 commented:
I have added a release note as this an API addition. I've added it to the `API changes` section although maybe it should go under `Additions` instead?

mattpap commented:
I would add a test or two for the ability to update bounds.

ianthomas23 commented:
> I would add a test or two for the ability to update bounds.

I am glad you suggested this, it revealed a bug where I wasn't handling vertical and horizontal colorbars the same.

I've rebased it to pick up the baseline image changes, added two update tests for horizontal/vertical EqHist/Linear color mappers, and fixed the bug.
